File: packages/gbnf-experiment/data/20260909T134433Z_3f62598c/ported_implementation/gbnf/utils/errors/build_error_position.py
from typing import List

MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3


def build_error_position(src: str, pos: int) -> List[str]:
    if src == "":
        return ["No input provided"]

    lines = src.split("\n")

    line_idx = 0
    # an empty line terminates the walk, matching the falsiness check of the reference
    while line_idx < len(lines) and lines[line_idx] and pos > len(lines[line_idx]) - 1:
        pos -= len(lines[line_idx])
        line_idx += 1

    lines_to_show: List[str] = []
    start = max(0, line_idx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1))
    for i in range(start, line_idx + 1):
        # when the position runs past the end of the input, the walk above stops on
        # a line that does not exist; the reference renders it as a blank line
        lines_to_show.append(lines[i] if i < len(lines) else "")

    return [
        *lines_to_show,
        " " * pos + "^",
    ]

File: proposed-projects/round-trip-experiment/reverse/20260909T134433Z_3f62598c/20260910T080424Z_8686d09e/ported_implementation/src/utils/errors/buildErrorPosition.ts
export const MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3;

export const buildErrorPosition = (src: string, pos: number): string[] => {
  if (src === '') {
    return ['No input provided'];
  }

  const lines = src.split('\n');

  let lineIdx = 0;
  // an empty line terminates the walk, matching a falsiness check on the line
  while (lineIdx < lines.length && lines[lineIdx] && pos > lines[lineIdx].length - 1) {
    pos -= lines[lineIdx].length;
    lineIdx += 1;
  }

  const linesToShow: string[] = [];
  const start = Math.max(0, lineIdx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1));
  for (let i = start; i <= lineIdx; i++) {
    // when the position runs past the end of the input, the walk above stops on
    // a line that does not exist; it is rendered as a blank line
    linesToShow.push(i < lines.length ? lines[i] : '');
  }

  return [
    ...linesToShow,
    `${' '.repeat(Math.max(0, pos))}^`,
  ];
};
